superset/migrations/shared/security_converge.py

```diff
@@ -0,0 +1,260 @@
+        .filter(Permission.name == permission_name)
+        .one_or_none()
+    )
+    if not new_permission:
+        new_permission = Permission(name=permission_name)
+        session.add(new_permission)
+    return new_permission
+
+
+def _add_permission_view(
+    session: Session, permission: Permission, view_menu: ViewMenu
+) -> PermissionView:
+    """
+    Check and add the new Permission View
+    """
+    new_pvm = (
+        session.query(PermissionView)
+        .filter(
+            PermissionView.view_menu_id == view_menu.id,
+            PermissionView.permission_id == permission.id,
+        )
+        .one_or_none()
+    )
+    if not new_pvm:
+        new_pvm = PermissionView(view_menu=view_menu, permission=permission)
+        session.add(new_pvm)
+    return new_pvm
+
+
+def _find_pvm(session: Session, view_name: str, permission_name: str) -> PermissionView:
+    return (
+        session.query(PermissionView)
+        .join(Permission)
+        .join(ViewMenu)
+        .filter(ViewMenu.name == view_name, Permission.name == permission_name)
+    ).one_or_none()
+
+
+def add_pvms(
+    session: Session, pvm_data: Dict[str, Tuple[str, ...]], commit: bool = False
+) -> None:
+    """
+    Checks if exists and adds new Permissions, Views and PermissionView's
+    """
+    for view_name, permissions in pvm_data.items():
+        # Check and add the new View
+        new_view = _add_view_menu(session, view_name)
+        for permission_name in permissions:
+            new_permission = _add_permission(session, permission_name)
+            # Check and add the new PVM
+            _add_permission_view(session, new_permission, new_view)
+    if commit:
+        session.commit()
+
+
+def _delete_old_permissions(
+    session: Session, pvm_map: Dict[PermissionView, List[PermissionView]]
+) -> None:
+    """
+    Delete old permissions:
+    - Delete the PermissionView
+    - Deletes the Permission if it's an orphan now
+    - Deletes the ViewMenu if it's an orphan now
+    """
+    # Delete old permissions
+    for old_pvm, new_pvms in pvm_map.items():
+        old_permission_name = old_pvm.permission.name
+        old_view_name = old_pvm.view_menu.name
+        print(f"Going to delete pvm: {old_pvm}")
+        session.delete(old_pvm)
+        pvms_with_permission = (
+            session.query(PermissionView)
+            .join(Permission)
+            .filter(Permission.name == old_permission_name)
+        ).first()
+        if not pvms_with_permission:
+            print(f"Going to delete permission: {old_pvm.permission}")
+            session.delete(old_pvm.permission)
+        pvms_with_view_menu = (
+            session.query(PermissionView)
+            .join(ViewMenu)
+            .filter(ViewMenu.name == old_view_name)
+        ).first()
+        if not pvms_with_view_menu:
+            print(f"Going to delete view_menu: {old_pvm.view_menu}")
+            session.delete(old_pvm.view_menu)
+
+
+def migrate_roles(
```

Comment:
I'd feel more comfortable if we wrote tests for these utility functions - they're pretty important and do have a fair amount of logic.

Reply:
Right you are, will do that